### PR TITLE
Rework Serializable class.

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong

The `Serializable` class was ready for some cleanup.

### How was it fixed

- replaced usage of `__setattr_` and `__getattr__` with metaclass mechanics and `property` based setters and getters.
- removed the `cls.exclude` API.
- changed `is_mutable` to be property based.
- removed mutation based `make_mutable` API in favor of a copy based API.
- implemented `__iter__`
- improved `__eq__` implementation to only return true on equality of *rich* values rather than serialized values.
- added new `obj[field_name]` and `obj[0]` based field lookups.
- removed `cls.fields` in favor of `cls._meta` which holds field names, sedes and some extra meta info.
- reworked `__init__` argument handling to be a bit more explicit in how it handles `args` and `kwargs`.

### CODE REVIEW

**Comment by pipermerriam** [comment]
@carver @gsalgado ready for high level preliminary review.  The `collections.namedtuple` idea didn't work out as planned due to the way that `namedtuple` works.  It doesn't really create a class that you can *properly* inherit from (or I was just doing it wrong because of metaclass wizardry).

File: rlp/sedes/serializable.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+        params = {
+            field: value
+            for field, value
+            in zip(cls._meta.field_names, values)
+            if exclude is None or field not in exclude
+        }
+        obj = cls(**dict(list(params.items()) + list(kwargs.items())))
+
+        # TODO: mutability
```
Comment (gsalgado) [inline rlp/sedes/serializable.py]: Is there anything really left todo here?

File: rlp/sedes/serializable.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        else:
+            return obj.as_immutable()
+
+    @classmethod
+    def exclude(cls, excluded_fields):
+        """Create a new sedes considering only a reduced set of fields."""
+        raise NotImplementedError("Not yet implemented")
+
+    @property
+    def is_immutable(self):
```
Comment (gsalgado) [inline rlp/sedes/serializable.py]: We probably don't need this as we can just inline the `not self.is_mutable`, no?

File: rlp/sedes/serializable.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+    def _get_field_attrs(self):
+        return self._meta.field_attrs
+
+
+class WritableSerializable(BaseSerializable):
+    is_mutable = True
+
+    _cached_rlp_value = None
+    _cached_rlp_key = None
```
Comment (gsalgado) [inline rlp/sedes/serializable.py]: There's something about having two separate attributes for this that makes me uncomfortable... Did you consider using a dict for that? Maybe `_cached_snapshot = {}`, or maybe a `_get_cached_rlp(snapshot_key)` method decorated with a `lru_cache`

File: rlp/sedes/serializable.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    def __iter__(self):
+        for field in self._meta.get_field_names():
+            yield getattr(self, field)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, int):
+            field = self._meta.field_names[idx]
+            return getattr(self, field)
+        else:
+            raise NotImplementedError("not implemented")
```
Comment (carver) [inline rlp/sedes/serializable.py]:
```py
elif isinstance(idx, str):
    return getattr(self, idx)
```

Seems like a straightforward and nice thing to do here.

File: rlp/sedes/serializable.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+    def __len__(self):
+        return len(self._meta.fields)
+
+    def __eq__(self, other):
+        if not hasattr(other, 'serialize'):
+            return False
+        return self.serialize(self) == other.serialize(other)
```
Comment (carver) [inline rlp/sedes/serializable.py]: I'm not sure this is enough. An rlp with type (bytes32, bytes32) and one with type (int, int) can be trivially made to be equal here, which feels wrong to me.

Maybe test the deserialized values?

Comment (pipermerriam) [reply rlp/sedes/serializable.py]: Agreed.  

> Note that this is the current way that `rlp.sedes.Serializable` determines equality: https://github.com/ethereum/pyrlp/blob/188abd6f536aaaf9689413650422991b7510f9a7/rlp/sedes/lists.py#L204-L208

I may go back to using a version of the following:

```python
return len(self) == len(other) and all(
    compare_eq(left, right)
    for left, right
    in zip(self, other)
)
```

Where `compare_eq` is a function which handles things like equality between `tuples` and `lists` with the same values which a naive `==` comparison will return `False`.

File: rlp/sedes/serializable.py (hunk 3 above)
Comment (pipermerriam) [reply rlp/sedes/serializable.py]: Alright, I've changed to use a class level cache.  Each subclass of `Serializable` now has it's own cache which is shared across instances of those objects as well as across both the mutable and immutable versions of those objects.

**Comment by pipermerriam** [comment]
Performance Numbers:

> Worth noting that this is *somewhat* of an apples/oranges comparison.  Serialization slowed down while deserialization speed up.  Overall, it's a speed gain assuming even distribution of calls to serialize/deserialize, but in reality, I don't know what that distribution is. 

cc @gsalgado 

#### This Branch

```
Block serializations / sec: 1103.41
Block deserializations / sec: 2197.05
TX serializations / sec: 15646.35
TX deserializations / sec: 27768.98
```

#### Master @ f30fe1e5c06749ccb5747552d3ef61a5e15626ee

```
Block serializations / sec: 1935.49
Block deserializations / sec: 774.44
TX serializations / sec: 25199.91
TX deserializations / sec: 13675.92
```

**Comment by gsalgado** [comment]
For block importing (i.e. `Chain.import_block()`), we seem to consistently spend more than twice as long in `rlp.decode` than in `rlp.encode_raw`, but that's with py-rlp 0.6.0

**Comment by pipermerriam** [comment]
@gsalgado @carver can I get a final round of 👁 👁

**Comment by pipermerriam** [comment]
@gsalgado looking at stats, it looks like we have 2x calls to `decode` than `encode` which means this could have a significant effect on the amount of time spend dealing width decoding rlp objects since this branch almost doubles the speed of decoding rlp `Serializable` instances.

File: rlp/codec.py [hunk 6]
```diff
@@ -218,7 +222,6 @@ def decode(rlp, sedes=None, strict=True, **kwargs):
         obj = sedes.deserialize(item, **kwargs)
         if hasattr(obj, '_cached_rlp'):
             obj._cached_rlp = rlp
-            assert not isinstance(obj, Serializable) or not obj.is_mutable()
```
Comment (pipermerriam) [inline rlp/codec.py]: This was removed in favor of having mutable `Serializable` instances automatically invalidate their cache.

File: tests/test_serializable.py [hunk 7]
```diff
@@ -29,185 +71,209 @@ class RLPType2(Serializable):
         (RLPType2, [RLPType1(8, b'a', (7, ''))], 'field2_2'),
     ),
 )
-def test_validation(rlptype, fields, exception_includes):
+def test_serializable_initialization_validation(rlptype, fields, exception_includes):
+    # TODO: expand tests to combine *args and **kwargs instantiation
```
Comment (carver) [inline tests/test_serializable.py]: Time to do?

File: tests/test_serializable.py [hunk 8]
```diff
@@ -29,185 +71,209 @@ class RLPType2(Serializable):
...
+def test_serializable_encoding_rlp_caching(rlp_obj):
+    assert rlp_obj._cached_rlp is None
+
+    # obj should start out without a cache
+    rlp_code = encode(rlp_obj)
+    assert rlp_obj._cached_rlp is None
+    assert rlp_obj.is_mutable is True
+
+    # cache should be populated now.
+    assert encode(rlp_obj, cache=True) == rlp_code
```
Comment (carver) [inline tests/test_serializable.py]: Does it make sense to turn on caching by default?

**Pull request merged**